solana/solana-ibc/programs/solana-ibc/src/lib.rs

```diff
@@ -139,7 +143,13 @@ pub mod solana_ibc {
         chain.check_staking_program(&current_ixn.program_id)?;
         let provable = storage::get_provable_from(&ctx.accounts.trie)?;
         chain.maybe_generate_block(&provable)?;
-        chain.set_stake((*ctx.accounts.sender.key).into(), amount)
+        // Setting the stake to 0 removes the validator from the next
+        // epoch so the validator key passed in argument should be the
+        // signer.
```

Comment:
I don’t understand this. Setting stake to 1 effectively removes validator from the set as well. Why is zero somehow special? Any change to stake should have the same kind of authorisation.

Reply:
so if the stake is below the minimum stake, validator is gonna be removed?

Reply:
i mean for delegation, it would be other user who would call right. Since we are calling from restaking program, as long as we make sure the logic is right, we can remove the above logic.

Reply:
I mean, even if the stake is above minimum, if there’s a lot of candidates, if one validator’s stake is set too low, they will be removed from the set of validators because there will be someone else with more stake.

My understanding was that it’s always restaking program which calls set_stake.  Restaking is holding the funds and tracking who delegates them.

Reply:
Yes, but previously we were just setting the stake on the account who is calling the set_stake, so i think we need an argument to pass the validator to which it is getting delegated.

Reply:
But what i am missing is updating the new stake with the previous one.

Reply:
> Yes, but previously we were just setting the stake on the account who is calling the set_stake, so i think we need an argument to pass the validator to which it is getting delegated.

Yes, I get that. But what I don’t get is why there’s a special case for `amount == 0`. It seems to me that all checking needs to be done in restaking program since it has all the state and accepts deposits.